src/visitors.rs

```diff
@@ -548,153 +556,299 @@ impl<'a, 'b: 'a, 'tcx: 'b> MirVisitor<'a, 'b, 'tcx> {
-            mir::Rvalue::Discriminant(place) => self.visit_discriminant(place),
             mir::Rvalue::Aggregate(aggregate_kinds, operands) => {
-                self.visit_aggregate(aggregate_kinds, operands)
+                self.visit_aggregate(path, aggregate_kinds, operands);
+            }
+        }
+    }
+
+    /// path = x (either a move or copy, depending on type of x), or path = constant.
+    fn visit_use(&mut self, path: Path, operand: &mir::Operand) {
+        debug!(
+            "default visit_use(path: {:?}, operand: {:?})",
+            path, operand
+        );
+        //let span = self.current_span;
+        match operand {
+            mir::Operand::Copy(place) => {
+                self.visit_used_copy(path, place);
             }
+            mir::Operand::Move(place) => {
+                self.visit_used_move(path, place);
+            }
+            mir::Operand::Constant(constant) => {
+                let mir::Constant {
+                    span,
+                    ty,
+                    user_ty,
+                    literal,
+                } = constant.borrow();
+                let const_value: AbstractValue =
+                    self.visit_constant(ty, *user_ty, literal).clone().into();
+                self.current_environment
+                    .update_value_at(path, const_value.with_provenance(*span));
+            }
+        };
+    }
+
+    /// For each (path', value) pair in the environment where path' is rooted in place,
+    /// add a (path'', value) pair to the environment where path'' is a copy of path re-rooted
+    /// with place.
+    fn visit_used_copy(&mut self, target_path: Path, place: &mir::Place) {
+        let rpath = self.visit_place(place);
+        let value_map = &self.current_environment.value_map;
+        for (path, value) in value_map
+            .iter()
+            .filter(|(p, _)| Self::path_is_rooted_by(*p, &rpath))
+        {
+            let qualified_path = Self::replace_root(&path, target_path.clone());
+            debug!("copying {:?} to {:?}", value, qualified_path);
+            value_map.insert(qualified_path, value.clone());
+        }
+    }
+
+    /// For each (path', value) pair in the environment where path' is rooted in place,
+    /// add a (path'', value) pair to the environment where path'' is a copy of path re-rooted
+    /// with place, and also remove the (path', value) pair from the environment.
+    fn visit_used_move(&mut self, target_path: Path, place: &mir::Place) {
+        let rpath = self.visit_place(place);
+        let value_map = &self.current_environment.value_map;
+        for (path, value) in value_map
+            .iter()
+            .filter(|(p, _)| Self::path_is_rooted_by(*p, &rpath))
+        {
+            let qualified_path = Self::replace_root(&path, target_path.clone());
+            debug!("moving {:?} to {:?}", value, qualified_path);
+            value_map.remove(&path);
```

Comment:
Bah, this is identical to `visit_used_copy` except for this. It's probably overkill just to eliminate code duplication, but implementing `move(self, Path, Value)` for `ValueMap` would fix this problem. Could be worth it if `value_map.move` would be useful elsewhere.

Reply:
Could be, I'll give it go.